File: NestSDKDemo/NestSDKDemo/ViewController.m
//
//  ViewController.m
//  NestSDKDemo
//
//  Created by Petro Akzhygitov on 7/01/16.
//  Copyright © 2016 Petro Akzhygitov. All rights reserved.
//

#import <NestSDK/NestSDKAuthorizationManager.h>
#import <NestSDK/NestSDKAuthorizationManagerAuthorizationResult.h>
#import <NestSDK/NestSDKConnectWithNestButton.h>
#import <NestSDK/NestSDKDataManager.h>
#import <NestSDK/NestSDKStructure.h>
#import <NestSDK/NestSDKSmokeCOAlarm.h>
#import <NestSDK/NestSDKAccessToken.h>
#import <NestSDK/NestSDKCamera.h>
#import <NestSDK/NestSDKThermostat.h>
#import "ViewController.h"

@interface ViewController ()

@property(nonatomic) NestSDKDataManager *dataManager;
@property(nonatomic) NSMutableArray *deviceObserverHandles;

@property(nonatomic) NestSDKObserverHandle structuresObserverHandle;

@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.dataManager = [[NestSDKDataManager alloc] init];
    self.deviceObserverHandles = [[NSMutableArray alloc] init];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    if ([NestSDKAccessToken currentAccessToken]) {
        [self observeStructures];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];

    [self removeObservers];
}

- (void)observeStructures {
    // Clean up previous observers
    [self removeObservers];

    // Start observing structures
    self.structuresObserverHandle = [self.dataManager observeStructuresWithBlock:^(NSArray <NestSDKStructure> *structuresArray, NSError *error) {
        [self logMessage:@"Structures updated!"];

        // Structure may change while observing, so remove all current device observers and then set all new ones
        [self removeDevicesObservers];

        // Cycle through all structures and set observers for all devices
        for (id <NestSDKStructure> structure in structuresArray) {
            [self logMessage:[NSString stringWithFormat:@"Found structure: %@!", structure.name]];

            [self observeThermostatsWithinStructure:structure];
            [self observeSmokeCOAlarmsWithinStructure:structure];
            [self observeCamerasWithinStructure:structure];
        }
    }];
}

- (void)observeThermostatsWithinStructure:(id <NestSDKStructure>)structure {
    for (NSString *thermostatId in structure.thermostats) {
        NestSDKObserverHandle handle = [self.dataManager observeThermostatWithId:thermostatId block:^(id <NestSDKThermostat> thermostat, NSError *error) {
            if (error) {
                [self logMessage:[NSString stringWithFormat:@"Error observing thermostat: %@", error]];

            } else {
                [self logMessage:[NSString stringWithFormat:@"Thermostat %@ updated! Current temperature in C: %.1f",
                                                            thermostat.name, thermostat.ambientTemperatureC]];
            }
        }];

        [self.deviceObserverHandles addObject:@(handle)];
    }
}

- (void)observeSmokeCOAlarmsWithinStructure:(id <NestSDKStructure>)structure {
    for (NSString *smokeCOAlarmId in structure.smokeCoAlarms) {
        NestSDKObserverHandle handle = [self.dataManager observeSmokeCOAlarmWithId:smokeCOAlarmId block:^(id <NestSDKSmokeCOAlarm> smokeCOAlarm, NSError *error) {
            if (error) {
                [self logMessage:[NSString stringWithFormat:@"Error observing smokeCOAlarm: %@", error]];

            } else {
                [self logMessage:[NSString stringWithFormat:@"smokeCOAlarm %@ updated! Current state: %lul",
                                  smokeCOAlarm.name, (unsigned long)smokeCOAlarm.coAlarmState]];
            }
        }];

        [self.deviceObserverHandles addObject:@(handle)];
    }
}

- (void)observeCamerasWithinStructure:(id <NestSDKStructure>)structure {
    for (NSString *cameraId in structure.cameras) {
        NestSDKObserverHandle handle = [self.dataManager observeCameraWithId:cameraId block:^(id <NestSDKCamera> camera, NSError *error) {
            if (error) {
                [self logMessage:[NSString stringWithFormat:@"Error observing camera: %@", error]];

            } else {
                [self logMessage:[NSString stringWithFormat:@"Camera %@ updated! Streaming state: %@",
                                                            camera.name, camera.isStreaming ? @"YES" : @"NO"]];
            }
        }];

        [self.deviceObserverHandles addObject:@(handle)];
    }
}

- (void)removeObservers {
    [self removeDevicesObservers];
    [self removeStructuresObservers];
}

- (void)removeDevicesObservers {
    for (NSNumber *handle in self.deviceObserverHandles) {
        [self.dataManager removeObserverWithHandle:handle.unsignedIntegerValue];
    }

    [self.deviceObserverHandles removeAllObjects];
}

- (void)removeStructuresObservers {
    [self.dataManager removeObserverWithHandle:self.structuresObserverHandle];
}

- (void)logMessage:(NSString *)message {
    self.nestInfoTextView.text = [self.nestInfoTextView.text stringByAppendingFormat:@"%@\n", message];
}

#pragma mark - NestSDKConnectWithNestButtonDelegate

- (void)connectWithNestButton:(NestSDKConnectWithNestButton *)connectWithNestButton
       didAuthorizeWithResult:(NestSDKAuthorizationManagerAuthorizationResult *)result error:(NSError *)error {

    if (error) {
        NSLog(@"Process error: %@", error);

    } else if (result.isCancelled) {
        NSLog(@"Cancelled");

    } else {
        NSLog(@"Authorized!");

        [self observeStructures];
    }
}

- (void)connectWithNestButtonDidUnauthorize:(NestSDKConnectWithNestButton *)connectWithNestButton {
    [self removeObservers];
}


@end

File: NestSDKDemoSwift/NestSDKDemoSwift/ViewController.swift
//
//  ViewController.swift
//  NestSDKDemoSwift
//
//  Created by Petro Akzhygitov on 31/01/16.
//  Copyright © 2016 Petro Akzhygitov. All rights reserved.
//

import UIKit
import NestSDK

class ViewController: UIViewController, NestSDKConnectWithNestButtonDelegate {

    @IBOutlet weak var connectWithNestButton: NestSDKConnectWithNestButton!
    @IBOutlet weak var nestInfoTextView: UITextView!

    var dataManager: NestSDKDataManager = NestSDKDataManager()
    var deviceObserverHandles: Array<NestSDKObserverHandle> = []

    var structuresObserverHandle: NestSDKObserverHandle = 0

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)

        // Check authorization
        if (NestSDKAccessToken.currentAccessToken() != nil) {
            observeStructures()
        }
    }

    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)

        // Clean up
        removeObservers()
    }

    func observeStructures() {
        // Clean up previous observers
        removeObservers()

        // Start observing structures
        structuresObserverHandle = dataManager.observeStructuresWithBlock({
            structuresArray, error in

            self.logMessage("Structures updated!")

            // Structure may change while observing, so remove all current device observers and then set all new ones
            self.removeDevicesObservers()

            // Iterate through all structures and set observers for all devices
            for structure in structuresArray as! [NestSDKStructure] {
                self.logMessage("Found structure: \(structure.name)!")

                self.observeThermostatsWithinStructure(structure)
                self.observeSmokeCOAlarmsWithinStructure(structure)
                self.observeCamerasWithinStructure(structure)
            }
        })
    }

    func observeThermostatsWithinStructure(structure: NestSDKStructure) {
        for thermostatId in structure.thermostats as! [String] {
            let handle = dataManager.observeThermostatWithId(thermostatId, block: {
                thermostat, error in

                if (error != nil) {
                    self.logMessage("Error observing thermostat: \(error)")

                } else {
                    self.logMessage("Thermostat \(thermostat.name) updated! Current temperature in C: \(thermostat.ambientTemperatureC)")
                }
            })

            deviceObserverHandles.append(handle)
        }
    }

    func observeSmokeCOAlarmsWithinStructure(structure: NestSDKStructure) {
        for smokeCOAlarmId in structure.smokeCoAlarms as! [String] {
            let handle = dataManager.observeSmokeCOAlarmWithId(smokeCOAlarmId, block: {
                smokeCOAlarm, error in

                if (error != nil) {
                    self.logMessage("Error observing smokeCOAlarm: \(error)")

                } else {
                    self.logMessage("smokeCOAlarm \(smokeCOAlarm.name) updated! Current state: \(smokeCOAlarm.coAlarmState.rawValue)")
                }
            })

            deviceObserverHandles.append(handle)
        }
    }

    func observeCamerasWithinStructure(structure: NestSDKStructure) {
        for cameraId in structure.cameras as! [String] {
            let handle = dataManager.observeCameraWithId(cameraId, block: {
                camera, error in

                if (error != nil) {
                    self.logMessage("Error observing camera: \(error)")

                } else {
                    self.logMessage("Camera \(camera.name) updated! Streaming state: \(camera.isStreaming)")
                }
            })

            deviceObserverHandles.append(handle)
        }
    }

    func removeObservers() {
        removeDevicesObservers();
        removeStructuresObservers();
    }

    func removeDevicesObservers() {
        for (_, handle) in deviceObserverHandles.enumerate() {
            dataManager.removeObserverWithHandle(handle);
        }

        deviceObserverHandles.removeAll()
    }

    func removeStructuresObservers() {
        dataManager.removeObserverWithHandle(structuresObserverHandle)
    }

    func logMessage(message: String) {
        nestInfoTextView.text = nestInfoTextView.text + "\(message)\n"
    }

    // MARK: NestSDKConnectWithNestButtonDelegate

    func connectWithNestButton(connectWithNestButton: NestSDKConnectWithNestButton!, didAuthorizeWithResult result: NestSDKAuthorizationManagerAuthorizationResult!, error: NSError!) {
        if (error != nil) {
            print("Process error: \(error)")

        } else if (result.isCancelled) {
            print("Cancelled")

        } else {
            print("Authorized!")

            observeStructures()
        }
    }

    func connectWithNestButtonDidUnauthorize(connectWithNestButton: NestSDKConnectWithNestButton!) {
        removeObservers()
    }
}

